# Goal: Extract the statistics from the NFL website and place it into a CSV file.

Background: The National Football League (NFL) is the official league for the sport of American Football. The league consists of 32 teams from various cities across the United States.  

I want to get each NFL team's statistics from the NFL website so I can see which factors influence an NFL team's performance the most. To do this, I decided to check out data from the official NFL website: https://www.nfl.com/stats/team-stats/offense/passing/2020/reg/all

The webpage does not offer a CSV form of this file at the moment. So I'll need to scrape the HTML file that comprises the webpage to extract the data I want.

I'll start by importing BeautifulSoup and requests, two modules that will help with web scraping, along with pandas and numpy, which will come in handy when I turn the extracted data into a dataframe.

Let's then use BeautifulSoup to read in the html file from the official NFL stats webpage:

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np

html_file = requests.get('https://www.nfl.com/stats/team-stats/offense/passing/2020/reg/all').text
soup = BeautifulSoup(html_file, 'lxml')

<img src="webpage_screenshots/webpage.PNG" class="img-responsive" alt="Screenshot of the NFL stats webpage" width="900" height="300">

Screenshot of the NFL stats webpage

Each NFL team's name is listed the HTML file as part of a 'div class = "d3-o-club-fullname"' tag. I can use BeautifulSoup's find() function to search the HTML file for an instance of this tag. Here I printed the results from that search:

In [2]:
team = soup.find('div', class_ = "d3-o-club-fullname").text
team.strip()

'Football Team'

I used the find() function above, which only returns the first search result. But using the find_all() function returns all of the search results, which means I can get the names of all of the NFL teams in the order they appear on the webpage:

In [3]:
team_names = soup.find_all('div', class_ = 'd3-o-club-fullname')
for i in range(len(team_names)):
    team_names[i] = team_names[i].text.strip()

team_names

['Football Team',
 'Buccaneers',
 'Seahawks',
 '49ers',
 'Chargers',
 'Steelers',
 'Cardinals',
 'Eagles',
 'Jets',
 'Giants',
 'Saints',
 'Patriots',
 'Vikings',
 'Dolphins',
 'Raiders',
 'Rams',
 'Chiefs',
 'Jaguars',
 'Colts',
 'Texans',
 'Titans',
 'Packers',
 'Lions',
 'Broncos',
 'Cowboys',
 'Browns',
 'Bengals',
 'Bears',
 'Panthers',
 'Bills',
 'Ravens',
 'Falcons']

When I make my dataframe of all the data from the NFL webpage, I want the dataframe's columns to match the columns on the webpage. So I'll extract the names of the columns on the webpage and store them in a list called col_names:

In [4]:
col_names = soup.find_all('th', scope = 'col')
#col_names = soup.find_all('thead')
for i in range(len(col_names)):
    col_names[i] = col_names[i].text.strip()

col_names

['Att',
 'Cmp',
 'Cmp %',
 'Yds/Att',
 'Pass Yds',
 'TD',
 'INT',
 'Rate',
 '1st',
 '1st%',
 '20+',
 '40+',
 'Lng',
 'Sck',
 'SckY']

<img src="webpage_screenshots/Columns.PNG" class="img-responsive" alt="Screenshot of the NFL stats webpage. The image contains a circle around the columns in the webpage's stats table." width="900" height="300">

Circled in the image above is the part of the NFL webpage where I draw the names of my columns from.

Great! Each team's data is loaded into the HTML webpage under the 'tr' tag. So I'll look inside one of these 'tr' tags to inspect its contents.

In [5]:
soup.find_all('tr')[1]

<tr>
<td scope="row" tabindex="0">
<div class="d3-o-club-info">
<div class="d3-o-club-logo">
<picture><!--[if IE 9]><video style="display: none; "><![endif]--><source media="(min-width:1024px)" srcset="https://static.www.nfl.com/t_q-best/league/api/clubs/logos/WAS"/><source media="(min-width:768px)" srcset="https://static.www.nfl.com/t_q-best/league/api/clubs/logos/WAS"/><source srcset="https://static.www.nfl.com/t_q-best/league/api/clubs/logos/WAS"/><!--[if IE 9]></video><![endif]--><img class="img-responsive" src="https://static.www.nfl.com/t_q-best/league/api/clubs/logos/WAS"/></picture>
</div>
<div class="d3-o-club-fullname">
            Football Team
            
          </div>
<div class="d3-o-club-shortname">
            Football Team
            
          </div>
</div>
</td>
<td>
                    601
                  </td>
<td>
                    389
                  </td>
<td>
                    64.7
                  </td>
<td>
                    6.3
              

<img src="webpage_screenshots/tr_tags.PNG" class="img-responsive" alt="Screenshot of the NFL stats webpage. The image contains circles around the rows for each team, which represent the tr tags in the HTML file" width="900" height="300">

Circled in the image above are the rows for the NFL teams in the webpage's stats table. These are represented by 'tr' tags in the HTML file.

From the HTML code above, I can now confirm that the name of the football team is indeed listed under the 'div class="d3-o-club-fullname"' tag. In this case, we're working with Washington's NFL team, which is officially named "Football Team". 

(Note: the webpage may be updated by the time you read this, and the table may be reordered in the future. In this case, you may see a different team name in the 'div' tag. The format, however, will most likely remain the same.)

Inside each 'td' tag is the stats for the specific NFL team I'm working with. Conveniently, the order of the stats matches the order of elements in col_names. So 601 represents 'Att', the number of pass attempts, 389 represents 'Cmp' the number of pass completions, etc.

The goal of my next block of code below is to get all of the stats for each NFL team and store them in a list.

I'll start by creating a set called teams. Since the webpage uses the 'tr' tag to denote each team, I will find_all() instances of the 'tr' tag. Each element of teams will contain the contents inside one of these 'tr' tags. I'll also create a list called data which I'll leave empty for now.

In [6]:
teams = soup.find_all('tr')
data = []

Great! Each element of teams will look a bit like the snippet of HTML for the Washington Football Team that we saw above. The only difference between each element will be the team name and the actual numbers that comprise that team's stats. The HTML formatting remains identical.

I'll loop through each team in teams. Each time I do so, I'll create a team_stats list that contains all of the stats for that team. The team_stats list starts out empty. For each team, I'll through each 'td' tag, convert the contents into a float format, then append it to the team_stats list.

At the end, the team_stats list will be full of stats for the particular team I'm working with. I'll take the team_stats list and append it to the 'data' list I created earlier.

To clarify:
The team_stats is a list of float point numbers that represents the stats of an __individual__ team.
The data list is a list of team_stats lists. The 'data' list contains the stats of __every__ NFL team.

The loop then continues onto the next team and repeats the process above.

In [7]:

for i in range(1, len(teams)):
    #i starts from 1 because teams[0] doesn't actually contain the stats of any particular team
    
    team_stats = []
    
    #go into the <td> section to find the stats
    stats = teams[i].find_all('td', scope=None)
    for stat in stats:
        stat = str(stat.text.strip())
        
        if 'T' in stat:
            stat = stat.replace('T', '')
            
        team_stats.append(float(stat))
    #these lines help parse the stats into a format the dataframe can read (namely floats)
    
    data.append(team_stats)

Here's what the team_stats list looks like for the final team we analyzed:

In [8]:
team_stats

[628.0,
 408.0,
 65.0,
 7.4,
 4620.0,
 27.0,
 11.0,
 93.9,
 243.0,
 38.7,
 59.0,
 8.0,
 63.0,
 41.0,
 257.0]

And here's what the first five elements of the data list look like:

In [9]:
data[:5]

[[601.0,
  389.0,
  64.7,
  6.3,
  3796.0,
  16.0,
  16.0,
  80.1,
  184.0,
  30.6,
  40.0,
  6.0,
  68.0,
  50.0,
  331.0],
 [626.0,
  410.0,
  65.5,
  7.6,
  4776.0,
  42.0,
  12.0,
  102.8,
  238.0,
  38.0,
  67.0,
  12.0,
  50.0,
  22.0,
  150.0],
 [563.0,
  388.0,
  68.9,
  7.5,
  4245.0,
  40.0,
  13.0,
  105.0,
  216.0,
  38.4,
  45.0,
  11.0,
  62.0,
  48.0,
  304.0],
 [570.0,
  371.0,
  65.1,
  7.6,
  4320.0,
  25.0,
  17.0,
  90.1,
  217.0,
  38.1,
  55.0,
  10.0,
  76.0,
  39.0,
  287.0],
 [627.0,
  413.0,
  65.9,
  7.3,
  4548.0,
  31.0,
  10.0,
  97.0,
  226.0,
  36.0,
  54.0,
  10.0,
  72.0,
  34.0,
  219.0]]

The 'data' list is rather hard to read right now. Let's turn it into a form that's easier to read by converting it into a dataframe. The new dataframe containing the data list will be called 'nfl'.

In [10]:
data = np.array(data)

In [11]:
nfl = pd.DataFrame(data = data)
nfl

0      1     2    3       4     5     6      7      8     9     10  \
0   601.0  389.0  64.7  6.3  3796.0  16.0  16.0   80.1  184.0  30.6  40.0   
1   626.0  410.0  65.5  7.6  4776.0  42.0  12.0  102.8  238.0  38.0  67.0   
2   563.0  388.0  68.9  7.5  4245.0  40.0  13.0  105.0  216.0  38.4  45.0   
3   570.0  371.0  65.1  7.6  4320.0  25.0  17.0   90.1  217.0  38.1  55.0   
4   627.0  413.0  65.9  7.3  4548.0  31.0  10.0   97.0  226.0  36.0  54.0   
5   656.0  428.0  65.2  6.3  4129.0  35.0  11.0   93.5  206.0  31.4  48.0   
6   575.0  387.0  67.3  7.1  4102.0  27.0  13.0   94.1  211.0  36.7  45.0   
7   598.0  334.0  55.9  6.2  3728.0  22.0  20.0   72.9  177.0  29.6  43.0   
8   499.0  292.0  58.5  6.2  3115.0  16.0  14.0   75.9  146.0  29.3  39.0   
9   517.0  321.0  62.1  6.5  3336.0  12.0  11.0   79.6  178.0  34.4  36.0   
10  522.0  370.0  70.9  7.6  3945.0  28.0   8.0  104.1  199.0  38.1  44.0   
11  440.0  283.0  64.3  7.1  3124.0  12.0  14.0   81.1  157.0  35.7  38.0   
12  516.0  349.0  67.6  8.3  4265.0  35.0  13.0  105.0  212.0  41.1  58.0   
13  559.0  370.0  66.2  7.0  3937.0  24.0  13.0   91.2  207.0  37.0  40.0   
14  551.0  369.0  67.0  8.0  4383.0  28.0  10.0  100.4  206.0  37.4  59.0   
15  590.0  392.0  66.4  7.1  4183.0  20.0  14.0   88.4  206.0  34.9  50.0   
16  630.0  420.0  66.7  7.9  5005.0  40.0   7.0  107.3  255.0  40.5  69.0   
17  616.0  387.0  62.8  6.4  3955.0  25.0  16.0   83.9  205.0  33.3  45.0   
18  552.0  371.0  67.2  7.6  4186.0  24.0  11.0   95.9  201.0  36.4  56.0   
19  546.0  383.0  70.1  8.9  4843.0  33.0   7.0  112.3  222.0  40.7  70.0   
20  485.0  316.0  65.2  7.9  3826.0  33.0   7.0  105.9  203.0  41.9  44.0   
21  526.0  372.0  70.7  8.2  4299.0  48.0   5.0  121.5  216.0  41.1  57.0   
22  582.0  374.0  64.3  7.6  4397.0  27.0  13.0   93.3  220.0  37.8  58.0   
23  556.0  317.0  57.0  6.6  3673.0  21.0  23.0   72.5  176.0  31.7  49.0   
24  639.0  413.0  64.6  7.1  4511.0  25.0  13.0   89.9  231.0  36.2  50.0   
25  501.0  315.0  62.9  7.4  3701.0  27.0   8.0   96.6  195.0  38.9  45.0   
26  581.0  372.0  64.0  6.5  3793.0  19.0  11.0   85.7  201.0  34.6  35.0   
27  614.0  402.0  65.5  6.4  3925.0  26.0  16.0   86.5  218.0  35.5  42.0   
28  550.0  373.0  67.8  7.5  4129.0  16.0  16.0   87.5  192.0  34.9  55.0   
29  596.0  410.0  68.8  8.0  4786.0  40.0  11.0  107.6  240.0  40.3  66.0   
30  406.0  257.0  63.3  7.2  2919.0  27.0  11.0   95.7  142.0  35.0  38.0   
31  628.0  408.0  65.0  7.4  4620.0  27.0  11.0   93.9  243.0  38.7  59.0   

      11    12    13     14  
0    6.0  68.0  50.0  331.0  
1   12.0  50.0  22.0  150.0  
2   11.0  62.0  48.0  304.0  
3   10.0  76.0  39.0  287.0  
4   10.0  72.0  34.0  219.0  
5    7.0  84.0  14.0  126.0  
6   14.0  80.0  29.0  186.0  
7    9.0  81.0  65.0  401.0  
8    6.0  69.0  43.0  319.0  
9    5.0  53.0  50.0  310.0  
10   9.0  52.0  29.0  187.0  
11   3.0  50.0  37.0  234.0  
12   7.0  71.0  39.0  256.0  
13   4.0  70.0  34.0  201.0  
14  12.0  85.0  28.0  166.0  
15   6.0  56.0  25.0  169.0  
16   8.0  75.0  24.0  151.0  
17   5.0  73.0  44.0  256.0  
18   8.0  55.0  21.0  133.0  
19  11.0  77.0  50.0  305.0  
20  13.0  75.0  25.0  173.0  
21  14.0  78.0  21.0  193.0  
22   8.0  73.0  42.0  293.0  
23  11.0  92.0  32.0  222.0  
24  11.0  69.0  44.0  350.0  
25   5.0  75.0  26.0  162.0  
26   5.0  72.0  48.0  345.0  
27   3.0  53.0  36.0  270.0  
28   9.0  75.0  36.0  241.0  
29  10.0  56.0  27.0  166.0  
30   5.0  70.0  32.0  180.0  
31   8.0  63.0  41.0  257.0

I can make this dataframe more readable my specifying the names of each row and column. Fortunately, I already have the column names saved under the col_names list. Each row represents an NFL team, and the ordering of teams in the team_names list just so happens to perfectly align with the ordering of each team's stats. So I can use the team_names list to name my rows:

In [12]:
nfl.index = team_names
nfl.columns = col_names
nfl

Att    Cmp  Cmp %  Yds/Att  Pass Yds    TD   INT   Rate  \
Football Team  601.0  389.0   64.7      6.3    3796.0  16.0  16.0   80.1   
Buccaneers     626.0  410.0   65.5      7.6    4776.0  42.0  12.0  102.8   
Seahawks       563.0  388.0   68.9      7.5    4245.0  40.0  13.0  105.0   
49ers          570.0  371.0   65.1      7.6    4320.0  25.0  17.0   90.1   
Chargers       627.0  413.0   65.9      7.3    4548.0  31.0  10.0   97.0   
Steelers       656.0  428.0   65.2      6.3    4129.0  35.0  11.0   93.5   
Cardinals      575.0  387.0   67.3      7.1    4102.0  27.0  13.0   94.1   
Eagles         598.0  334.0   55.9      6.2    3728.0  22.0  20.0   72.9   
Jets           499.0  292.0   58.5      6.2    3115.0  16.0  14.0   75.9   
Giants         517.0  321.0   62.1      6.5    3336.0  12.0  11.0   79.6   
Saints         522.0  370.0   70.9      7.6    3945.0  28.0   8.0  104.1   
Patriots       440.0  283.0   64.3      7.1    3124.0  12.0  14.0   81.1   
Vikings        516.0  349.0   67.6      8.3    4265.0  35.0  13.0  105.0   
Dolphins       559.0  370.0   66.2      7.0    3937.0  24.0  13.0   91.2   
Raiders        551.0  369.0   67.0      8.0    4383.0  28.0  10.0  100.4   
Rams           590.0  392.0   66.4      7.1    4183.0  20.0  14.0   88.4   
Chiefs         630.0  420.0   66.7      7.9    5005.0  40.0   7.0  107.3   
Jaguars        616.0  387.0   62.8      6.4    3955.0  25.0  16.0   83.9   
Colts          552.0  371.0   67.2      7.6    4186.0  24.0  11.0   95.9   
Texans         546.0  383.0   70.1      8.9    4843.0  33.0   7.0  112.3   
Titans         485.0  316.0   65.2      7.9    3826.0  33.0   7.0  105.9   
Packers        526.0  372.0   70.7      8.2    4299.0  48.0   5.0  121.5   
Lions          582.0  374.0   64.3      7.6    4397.0  27.0  13.0   93.3   
Broncos        556.0  317.0   57.0      6.6    3673.0  21.0  23.0   72.5   
Cowboys        639.0  413.0   64.6      7.1    4511.0  25.0  13.0   89.9   
Browns         501.0  315.0   62.9      7.4    3701.0  27.0   8.0   96.6   
Bengals        581.0  372.0   64.0      6.5    3793.0  19.0  11.0   85.7   
Bears          614.0  402.0   65.5      6.4    3925.0  26.0  16.0   86.5   
Panthers       550.0  373.0   67.8      7.5    4129.0  16.0  16.0   87.5   
Bills          596.0  410.0   68.8      8.0    4786.0  40.0  11.0  107.6   
Ravens         406.0  257.0   63.3      7.2    2919.0  27.0  11.0   95.7   
Falcons        628.0  408.0   65.0      7.4    4620.0  27.0  11.0   93.9   

                 1st  1st%   20+   40+   Lng   Sck   SckY  
Football Team  184.0  30.6  40.0   6.0  68.0  50.0  331.0  
Buccaneers     238.0  38.0  67.0  12.0  50.0  22.0  150.0  
Seahawks       216.0  38.4  45.0  11.0  62.0  48.0  304.0  
49ers          217.0  38.1  55.0  10.0  76.0  39.0  287.0  
Chargers       226.0  36.0  54.0  10.0  72.0  34.0  219.0  
Steelers       206.0  31.4  48.0   7.0  84.0  14.0  126.0  
Cardinals      211.0  36.7  45.0  14.0  80.0  29.0  186.0  
Eagles         177.0  29.6  43.0   9.0  81.0  65.0  401.0  
Jets           146.0  29.3  39.0   6.0  69.0  43.0  319.0  
Giants         178.0  34.4  36.0   5.0  53.0  50.0  310.0  
Saints         199.0  38.1  44.0   9.0  52.0  29.0  187.0  
Patriots       157.0  35.7  38.0   3.0  50.0  37.0  234.0  
Vikings        212.0  41.1  58.0   7.0  71.0  39.0  256.0  
Dolphins       207.0  37.0  40.0   4.0  70.0  34.0  201.0  
Raiders        206.0  37.4  59.0  12.0  85.0  28.0  166.0  
Rams           206.0  34.9  50.0   6.0  56.0  25.0  169.0  
Chiefs         255.0  40.5  69.0   8.0  75.0  24.0  151.0  
Jaguars        205.0  33.3  45.0   5.0  73.0  44.0  256.0  
Colts          201.0  36.4  56.0   8.0  55.0  21.0  133.0  
Texans         222.0  40.7  70.0  11.0  77.0  50.0  305.0  
Titans         203.0  41.9  44.0  13.0  75.0  25.0  173.0  
Packers        216.0  41.1  57.0  14.0  78.0  21.0  193.0  
Lions          220.0  37.8  58.0   8.0  73.0  42.0  293.0  
Broncos        176.0  31.7  49.0  11.0  92.0  32.0  222.0  
Cowboys 

Ahhh, there we go! Finally, a dataframe that is clear and simple is to read!

My final step is to save the dataframe I just made into a csv file. I chose to call my csv file 'nflstats.csv'.
After running the following line of code, a file called 'nflstats.csv' containing the stats for each NFL team will show up in the same folder as this notebook file.

In [13]:
nfl.to_csv(r'nflstats.csv')

Mission complete! The stats are now prepared, cleaned, and ready for data analysis. Thank you for reading until the end of this notebook. As always, if you have any suggestions, feel free to let me know!